docs/application-tutorial/game_of_life.py

```diff
@@ -0,0 +1,511 @@
+
+weights_method_4b = np.array([[1, 1, 1], [1, 0, 1], [1, 1, 1]])
+table_next_cell_4b_a = [i - 1 if i in [2, 3] else 0 for i in range(9)]
+table_next_cell_4b_b = [i in [2, 3] for i in range(4)]
+
+weights_method_5b = np.array([[1, 1, 1], [1, 9, 1], [1, 1, 1]])
+table_next_cell_5b = [i in [3, 9 + 2, 9 + 3] for i in range(18)]
+
+weights_method_basic = np.array([[1, 1, 1], [1, 0, 1], [1, 1, 1]])
+table_next_cell_basic_a = [i in [3] for i in range(9)]
+table_next_cell_basic_b = [i in [2, 3] for i in range(9)]
+
+
+# CP tables
+table_cp_next_cell_3b_a = fhe.LookupTable(list(map(int, table_next_cell_3b_a)))
+table_cp_next_cell_3b_b = fhe.LookupTable(list(map(int, table_next_cell_3b_b)))
+table_cp_next_cell_3b_c = fhe.LookupTable(list(map(int, table_next_cell_3b_c)))
+
+table_cp_next_cell_4b_a = fhe.LookupTable(list(map(int, table_next_cell_4b_a)))
+table_cp_next_cell_4b_b = fhe.LookupTable(list(map(int, table_next_cell_4b_b)))
+
+table_cp_next_cell_5b = fhe.LookupTable(list(map(int, table_next_cell_5b)))
+
+table_cp_next_cell_basic_a = fhe.LookupTable(list(map(int, table_next_cell_basic_a)))
+table_cp_next_cell_basic_b = fhe.LookupTable(list(map(int, table_next_cell_basic_b)))
+
+# Function to workaround the miss of padding in CP
+def by_hand_padding(original_grid, res):
+    padded_res = fhe.zeros(original_grid.shape)
+
+    original_grid_shape = original_grid.shape
+    padded_res[0, 0, 1 : original_grid_shape[2] - 1, 1 : original_grid_shape[3] - 1] = res
+
+    assert original_grid.shape == padded_res.shape
+
+    return padded_res
+
+
+# Function for Game of Life
+@fhe.compiler({"grid": "encrypted"})
+def update_grid_method_3b(grid):
+
+    # Method which uses two first PBS of 3 bits and a third PBS of 2 bits
+
+    # This is to workaround the fact that we have no pad option in fhe.conv
+    do_padded_fix = True
+
+    # Compute the sum of 7 elements
+    convoluted_grid = fhe.conv(
+        grid,
+        weights_method_3b_a.reshape(1, 1, *weights_method_3b_a.shape),
+        strides=(1, 1),
+        dilations=(1, 1),
+        group=1,
+        pads=(1, 1, 1, 1) if not do_padded_fix else (0, 0, 0, 0),
+    )
+
+    if do_padded_fix:
+        convoluted_grid = by_hand_padding(grid, convoluted_grid)
+
+    # Apply a PBS: input in [0, 7], output in [0, 4]
+    grid_a = table_cp_next_cell_3b_a[convoluted_grid]
+
+    # Add the 8th one: output is in [0, 5]
+    convoluted_grid = fhe.conv(
+        grid,
+        weights_method_3b_b.reshape(1, 1, *weights_method_3b_b.shape),
+        strides=(1, 1),
+        dilations=(1, 1),
+        group=1,
+        pads=(1, 1, 1, 1) if not do_padded_fix else (0, 0, 0, 0),
+    )
+
+    if do_padded_fix:
+        convoluted_grid = by_hand_padding(grid, convoluted_grid)
+
+    grid_b = grid_a + convoluted_grid
+
+    # Apply a PBS: input in [0, 5], output in [0, 4]
+    grid_c = table_cp_next_cell_3b_b[grid_b]
+
+    # Add center
+    grid = grid_c + grid
+
+    # And a last PBS
```

Comment:
Maybe show input and output range just like the other PBSes above.

Reply:
sure for both, done